gateway/src/kaazing/gateway/internal/ws/WebSocketExtensionHandler.js

```diff
@@ -0,0 +1,116 @@
+/**
+ * Copyright (c) 2007-2014 Kaazing Corporation. All rights reserved.
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+// It is injected in the message processing pipeline. The extension handler
+// delegates the control to the negotiated extensions
+var WebSocketExtensionHandler = (function(){
+
+    var WebSocketExtensionHandler = function() {
+        this._negotiatedExtensionSpis = [];
+    }
+
+    var $prototype = WebSocketExtensionHandler.prototype = new WebSocketHandlerAdapter();
+
+    $prototype.handleConnectionOpened = function(channel, protocol) {
+        var negotiatedExtensions = channel._negotiatedExtensions;
+        if (negotiatedExtensions) {
+            for (var i = 0; i < negotiatedExtensions.length; i++) {
+                var extensionElements = negotiatedExtensions[i].split(";");
+                var extensionName = extensionElements[0].replace(/^\s+|\s+$/g,"");
+                var extensionParameter = null;
+                var extensionInfo = WebSocketExtensionSpi.getRegisteredExtensionInfo(extensionName);
```

Comment:
Can we have `WebSocketExtensionFactory.get(extensionName)` return `WebSocketFactorySpi`? Then, the code in this function become very clean like this:

```
var extensionFactorySpi = WebSocketExtensionFactory.getFactory(extensionName);
....
var extensionSpi = extensionFactorySpi.create(extensionParameter);